main.go

```diff
@@ -14,191 +14,336 @@ import (
+func (e *PeriodicExporter) scrapeSlaves() {
+	hostnames := e.lockSlaves(func() []string {
+		return e.slaves.hostnames
+	})
+
+	totalHostnames := len(hostnames)
+	glog.V(6).Infof("active slaves: %d", totalHostnames)
+
+	dispatchCh := make(chan string)
+	ch := make(chan prometheus.Gauge)
+	go e.setMetrics(ch)
+
+	poolSize := concurrentFetch
+	if totalHostnames < concurrentFetch {
+		poolSize = totalHostnames
+	}
+
+	glog.V(6).Infof("creating fetch pool of size %d", poolSize)
+
+	var wg sync.WaitGroup
+	wg.Add(poolSize)
+	for i := 0; i < poolSize; i++ {
+		go e.fetch(dispatchCh, ch, &wg)
+	}
+
+	for _, host := range hostnames {
+		dispatchCh <- host
+	}
+	close(dispatchCh)
+
+	wg.Wait()
+	close(ch)
 }
 
-func main() {
-	flag.Parse()
+func (e *PeriodicExporter) updateSlaves() {
+	glog.V(6).Info("discovering slaves...")
+
+	// This will redirect us to the elected mesos master
+	redirectURL := fmt.Sprintf("%s:5050/master/redirect", e.opts.masterURL)
+	rReq, _ := http.NewRequest("GET", redirectURL, nil)
+
+	tr := http.Transport{}
+	rresp, err := tr.RoundTrip(rReq)
+	if err != nil {
+		glog.Warningf("GET %s failed. Error: %s", redirectURL, err)
+		return
+	}
+	defer rresp.Body.Close()
 
-	if len(*configFile) < 2 {
-		fmt.Printf("Error: Flag -config.file is required.\n")
-		flag.PrintDefaults()
-		os.Exit(1)
+	// This will/should return http://master.ip:5050
+	masterLoc := rresp.Header.Get("Location")
+	if masterLoc == "" {
+		glog.Warningf("%d response missing Location header", rresp.StatusCode)
+		return
 	}
 
-	config, err := newConfig()
+	glog.V(6).Infof("current elected master at: %s", masterLoc)
+
+	// Find all active slaves
+	stateURL := fmt.Sprintf("%s/master/state.json", masterLoc)
+	resp, err := http.Get(stateURL)
 	if err != nil {
-		fmt.Printf("Configuration error: %s\n", err)
-		os.Exit(2)
+		glog.Warningf("GET %s failed. Error: %s", stateURL, err)
+		return
 	}
+	defer resp.Body.Close()
 
-	go func() {
-		http.Handle(*metricsPath, prometheus.Handler())
-		http.HandleFunc("/status", func(w http.ResponseWriter, r *http.Request) {
-			fmt.Fprintf(w, "OK")
-		})
-		http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-			http.Redirect(w, r, *metricsPath, http.StatusMovedPermanently)
-		})
+	type slave struct {
+		Active   bool   `json:"active"`
+		Hostname string `json:"hostname"`
+	}
 
-		processRequests(config)
+	var req struct {
+		Slaves []*slave `json:"slaves"`
+	}
 
-		glog.Infof("Starting up on %s\n", config.Port)
-		err := http.ListenAndServe(":"+config.Port, nil)
-		if err != nil {
-			glog.Fatal(err)
+	if err := json.NewDecoder(resp.Body).Decode(&req); err != nil {
+		glog.Warningf("failed to deserialize request: %s", err)
+		return
+	}
+
+	var slaveHostnames []string
+	for _, slave := range req.Slaves {
+		if slave.Active {
+			slaveHostnames = append(slaveHostnames, slave.Hostname)
 		}
-	}()
+	}
+
+	glog.V(6).Infof("%d slaves discovered", len(slaveHostnames))
+
+	e.slaves.hostnames = e.lockSlaves(func() []string {
```

Comment:
No we lock before we return and unlock after we have returned? 
But I see what you mean, will remove the helper function all together. Doesn't really add any value